### PR TITLE
ENH Allow multiple backtick variables in a single value

### PR DESCRIPTION
**Pull request by emteknetnz**
https://github.com/silverstripe/silverstripe-framework/issues/10305

Required for the config update in https://github.com/silverstripe/cwp-core/pull/107

### CODE REVIEW

**Comment by maxime-rainville** [comment]
Can you clarify what's the use case this enables that wasn't possible before ... if only for posterity.

This also looks like something that should be documented in the changelog and/or doc?

**Comment by emteknetnz** [comment]
Required so that we can apply this config in the [cwp-core PR](https://github.com/silverstripe/cwp-core/pull/107)
```
- proxy: '`SS_OUTBOUND_PROXY`:`SS_OUTBOUND_PROXY_PORT`'
```

**Comment by michalkleiner** [comment]
I'll play the devil's advocate here this time...

Should be targetting `4` if it's an enhancement?
If we need it to fix `4.11` release (or CWP 2.11 depending on this in 4.11) it should be called a bugfix for a _bug_ where we would say we expected multiple variables to be evaluated and that wasn't working.

**Comment by emteknetnz** [comment]
Yes normally I'd do `4`, it's just there hasn't been stable to on 4.11 yet, only 4.11.0-beta1.  So it's correct to target 4.11 here.

File: src/Core/Injector/Injector.php [hunk 1]
```diff
@@ -524,16 +524,28 @@ public function convertServiceProperty($value)
...
-                $value = $envValue;
-            } elseif (defined($matches['name'] ?? '')) {
-                $value = constant($matches['name'] ?? '');
-            } else {
-                $value = null;
+                $val = $envValue;
+            } elseif (defined($name)) {
+                $val = constant($name);
+            }
+            $value = str_replace("`$name`", $val, $value);
```
Comment (GuySartorelli) [inline src/Core/Injector/Injector.php:540]: If there is no value for `$val` here is there potential here that we could be breaking someone's config? e.g. where previously something like ``'some value `with a tick` here'`` would have just been used as is, now it will be turned into `some value  here`.

File: tests/php/Core/Injector/InjectorTest.php [hunk 2]
```diff
@@ -815,6 +816,31 @@ public function testConvertServicePropertyOnCreate()
...
+
+    public function provideConvertServicePropertyBackTicks()
+    {
+        return [
+            ['`INJECTOR_TEST_CSP_A`', 'ABC'],
+            ['`INJECTOR_TEST_CSP_A`:`INJECTOR_TEST_CSP_B`', 'ABC:DEF'],
+            ['`INJECTOR_TEST_CSP_A`:`INJECTOR_TEST_CSP_C`', 'ABC:'],
+            ['`INJECTOR_TEST_CSP_C`', null],
+            ['lorem `INJECTOR_TEST_CSP_A` ipsum', 'lorem ABC ipsum'],
+            ['lorem `INJECTOR_TEST_CSP_A`:`INJECTOR_TEST_CSP_B` ipsum', 'lorem ABC:DEF ipsum'],
```
[the remark this answers is not in the document]
Comment (emteknetnz) [reply tests/php/Core/Injector/InjectorTest.php:838]: Added

File: src/Core/Injector/Injector.php (hunk 1 above)
Comment (emteknetnz) [reply src/Core/Injector/Injector.php:540]: I don't think we need to consider this edge case, because in Silverstripe backticks are used to denote string substituting in an environment variable or constant

The proper way to deal with this would be to have specific logic for escaping backticks to prevent string substitution, though I be surprised if there's a single example of someone doing this in the wild, so I don't think we need to solve this right now

Comment (emteknetnz) [reply src/Core/Injector/Injector.php:540]: Doing a regex search in vscode for a sink project
`['"][^`]+``
`**/*.yml` (includes ALL yml in the vendor folder)

The only matches are in `vendor/silverstripe/framework/.upgrade.yml` e.g.
```
      message: 'Session can not be accessed statically and `get_all()` is now called `getAll()'
    'SilverStripe\Control\Session::clear_all()':
```

So, perhaps there is a slightly concern we'd chop a few strings?  If this is a concern then we'd need to reconsider this whole thing

Comment (GuySartorelli) [reply src/Core/Injector/Injector.php:540]: > in Silverstripe backticks are used to denote string substituting in an environment variable or constant

This is only true (before this PR) of a value that entirely consists of an environment variable or constant - the regex used (``^`(?<name>[^`]+)`$``) means that only if there was a backtick at the beginning and the end of the value would it be used in this way. We're changing that with this PR.

> I be surprised if there's a single example of someone doing this in the wild

I wouldn't be surprised at all to see someone throwing backticks in their config for any number of reasons (javascript template literal that they'll later inject somewhere in a templated javascript file or `Requirements::customScript()` call, surrounding table or column names in something intended for use in a direct SQL query, or for some arbitrary custom purpose)
I don't think we can discount this outright.

Comment (GuySartorelli) [reply src/Core/Injector/Injector.php:540]: We can avoid this by _not_ replacing these with an empty value in situations where the value isn't entirely an env var or constant. 

e.g.
```yml
SomeClass:
  some_config: '`THIS_WILL_BE_EMPTY`'
  some_config: '`this will be empty`'
  some_config: 'the following `will not` be empty'
```

Comment (emteknetnz) [reply src/Core/Injector/Injector.php]: I've updated the logic so that `$value` must start and end with a backtick, though you can have multiple things that get substituted within `$value

This should make this change backwards compatible

**Comment by GuySartorelli** [comment]
What's the expected behaviour if someone has config that looks like ``'`something`somethingelse`'``?

File: tests/php/Core/Injector/InjectorTest.php [hunk 3]
```diff
@@ -815,6 +816,28 @@ public function testConvertServicePropertyOnCreate()
...
+        Environment::setEnv('INJECTOR_TEST_CSP_B', 'DEF');
+        $actual = Injector::inst()->convertServiceProperty($value);
+        $this->assertSame($expected, $actual);
+    }
+
+    public function provideConvertServicePropertyBackTicks()
+    {
+        return [
+            ['`INJECTOR_TEST_CSP_A`', 'ABC'],
+            ['`INJECTOR_TEST_CSP_A`:`INJECTOR_TEST_CSP_B`', 'ABC:DEF'],
```
Comment (GuySartorelli) [inline tests/php/Core/Injector/InjectorTest.php:834]:
```suggestion
            ['`INJECTOR_TEST_CSP_A`:`INJECTOR_TEST_CSP_B`', 'ABC:DEF'],
            ['`INJECTOR_TEST_CSP_A` some text `INJECTOR_TEST_CSP_B`', 'ABC some text DEF'],
            ['`INJECTOR_TEST_CSP_A``INJECTOR_TEST_CSP_B`', 'ABCDEF'],
```
Adding in an example with no characters between, and with some non-colon characters between the two vars.

Comment (GuySartorelli) [reply tests/php/Core/Injector/InjectorTest.php:834]: What would be the intended outcome of ```'`INJECTOR_TEST_CSP_C``INJECTOR_TEST_CSP_D`'```? I'm guessing null but we should probably test that case as well.

Comment (GuySartorelli) [reply tests/php/Core/Injector/InjectorTest.php:834]: Can you please also add a test case with three variables? Just to be sure it's not just grabbing the first and last.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
The actual functionality looks good. Just some additional test cases to catch edge cases.

File: tests/php/Core/Injector/InjectorTest.php (hunk 3 above)
Comment (emteknetnz) [reply tests/php/Core/Injector/InjectorTest.php:834]: Updated

**Review by GuySartorelli — APPROVED** [review]
Looks good. Merge on green.
Once merged, kick off the tests again on silverstripe/cwp-core#107
